contrib/babelfishpg_common/src/datetime.c

```diff
@@ -797,4 +804,548 @@ datetime_to_numeric(PG_FUNCTION_ARGS)
+		type = UNITS;
+		val = DTK_DOY;
+	}
+	if(strncmp(lowunits, "nanosecond", 11) == 0) {
+		type = UNITS;
+		val = DTK_NANO;
+	}
+	if(strncmp(lowunits, "weekday", 7) == 0) {
+		type = UNITS;
+		val = DTK_DAY;
+	}
+
+
+	if(type == UNITS) {
+		switch(val) {
+			case DTK_YEAR:
+				if(dttype == TIME) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type time.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, num, 0, 0, 0, 0, 0, 0);
+				break;
+			case DTK_QUARTER:
+				if(dttype == TIME) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type time.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, num * 3, 0, 0, 0, 0, 0);
+				break;
+			case DTK_MONTH:
+				if(dttype == TIME) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type time.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, num, 0, 0, 0, 0, 0);
+				break;
+			case DTK_WEEK:
+				if(dttype == TIME) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type time.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, num, 0, 0, 0, 0);
+				break;
+			case DTK_DAY:
+			case DTK_DOY:
+				if(dttype == TIME) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type time.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, num, 0, 0, 0);
+				break;
+			case DTK_HOUR:
+				if(dttype == DATE) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type date.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, num, 0, 0);
+				break;
+			case DTK_MINUTE:
+				if(dttype == DATE) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type date.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, num, 0);
+				break;
+			case DTK_SECOND:
+				if(dttype == DATE) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type date.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum(num));
+				break;
+			case DTK_MILLISEC:
+				if(dttype == DATE) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type date.",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum((float) num * 0.001));
+				break;
+			case DTK_MICROSEC:
+				if(dttype == SMALLDATETIME || dttype == DATETIME || dttype == DATE) {
+					elog(ERROR, "The datepart %s is not supported by date function dateadd for data type %s.",
```

Comment:
Would you reformat this part of code, to make sure all this error out in the same place and the err msg is also not the same cause it has an additional 'The' in the msg body.